### DISCUSSION TITLE
NFQUEUE doesn't work for routed traffic inside logical bridge (br-lan) on OpenWrt with nftables

### DISCUSSION THREAD

**Discussion by ValdikSS** [bol-van/zapret#262] [Announcements]
_…or, more precisely, `connbytes` doesn't match the traffic routed inside the bridge._

For some reason, nftables ip/ip6 family (not bridge family!) rules does not work when I'm trying to use `connbytes` matcher for _routing (not switching), when the routing occurs inside the bridge_, which doesn't allow `nfqws` to work out of the box.

![Network connection diagram](https://github.com/user-attachments/assets/5ff9bf88-3efa-439a-9f7c-145cf038dece)

```
chain zapret_lan_hook {
  type filter hook forward priority mangle;

  ip daddr != {0.0.0.0/8, 10.0.0.0/8, 192.168.0.0/16, 172.16.0.0/12, 100.64.0.0/10, 169.254.0.0/16, 224.0.0.0/3, 255.255.255.255/32} \
  tcp dport {80, 443} ct original packets lt 8 counter queue flags bypass to 200 comment "zapret IPv4 443 tcp"
}
```

does NOT work for some reason.

Turns out nftables skips conntrack after the NEW state even for routing, when it doesn't leave logical bridge. I had to install _iptables_ bridging module and enable _iptables_ bridge filtering (even if I don't need to filter bridge traffic, my traffic is routed).

**Solution:**

```
opkg install kmod-br-netfilter
sysctl net.bridge.bridge-nf-call-iptables=1
```

It fixed the issue, now I can use connbytes inside _nftables_ rules for the routed traffic inside br-lan.

P.S. I have all offloading disabled. OpenWrt 23.05.4. Honestly, this sounds like a nftables bug/deficiency.

**Comment by LiviaMedeiros** [comment]
Confirming the issue on raspberry pi, kernel v6.6.31.
Confirming the fix (`modprobe br_netfilter`) as well.

`br_netfilter` was removed from the bridge core with https://github.com/torvalds/linux/commit/34666d467cbf1e2e3c7bb15a63eccfb582cdd71f for performance reasons. Most people want hookless bridges with minimum overhead, so it makes sense to have it disabled by default.

However, i don't see that `br-netfilter` requires *iptables* here. It works just fine even after yeeting the legacy ip_tables module from kernel. The [tests](https://github.com/torvalds/linux/blob/d5d547aa7b51467b15d9caa86b116f8c2507c72a/tools/testing/selftests/net/netfilter/br_netfilter_queue.sh) are also implying bare nftables. Perhaps `bridge-nf-call-iptables` attribute sounds misleading?

**Comment by bol-van** [comment]
the simplest way - setup routing not bridge
separate ip subnet for clients behind openwrt device
static route on modem if possible
if not use masquerade

**Reply by hamedsbt** [reply]
Separation ip subnet for clients behind openwrt device:
Cilent (192.168.0.2) --> Raspberry (192.168.0.1 <> 192.168.1.2) --> Modem (192.168.1.1) --> Internet

- OpenWRT is installed on Raspberry & Modem.
- Client can ping the Modem and the Internet.

Same issue, Zapret didn't catch any packets! 

Raspberry /etc/config/network :
```
config interface 'loopback'
	option device 'lo'
	option proto 'static'
	option ipaddr '127.0.0.1'
	option netmask '255.0.0.0'

config globals 'globals'
	option ula_prefix 'fdd4:93bb:b9aa::/48'

config interface 'lan'
	option device 'eth0'
	option proto 'static'
	option ipaddr '192.168.0.1'
	option netmask '255.255.255.0'
	option ipv6 '0'
	list dns '127.0.0.1'
	option metric '1'

config interface 'adsl'
	option device 'eth0'
	option proto 'static'
	option ipaddr '192.168.1.2'
	option netmask '255.255.255.0'
	option gateway '192.168.1.1'
	list dns '127.0.0.1'
	option metric '2'
```

/etc/init.d/zapret list_table

```
table inet zapret {
	set zapret {
		type ipv4_addr
		size 522288
		flags interval
		auto-merge
	}

	set ipban {
		type ipv4_addr
		size 522288
		flags interval
		auto-merge
	}

	set nozapret {
		type ipv4_addr
		size 65536
		flags interval
		auto-merge
	}

	set lanif {
		type ifname
	}

	set wanif {
		type ifname
	}

	set wanif6 {
		type ifname
	}

	map link_local {
		type ifname : ipv6_addr
	}

	chain dnat_output {
		type nat hook output priority dstnat - 1; policy accept;
	}

	chain dnat_pre {
		type nat hook prerouting priority dstnat - 1; policy accept;
	}

	chain forward {
		type filter hook forward priority filter - 1; policy accept;
	}

	chain input {
		type filter hook input priority filter - 1; policy accept;
		iif != "lo" jump localnet_protect
	}

	chain flow_offload {
		tcp dport { 80, 443 } ct original packets 1-9 ip daddr != @nozapret return comment "direct flow offloading exemption"
		udp dport 443 ct original packets 1-9 ip daddr != @nozapret return comment "direct flow offloading exemption"
	}

	chain localnet_protect {
		ip daddr 127.0.0.127 return comment "route_localnet allow access to tpws"
		ip daddr 127.0.0.0/8 drop comment "route_localnet remote access protection"
	}

	chain postrouting {
	}

	chain postrouting_hook {
		type filter hook postrouting priority srcnat - 1; policy accept;
		meta mark & 0x40000000 == 0x00000000 jump postrouting
	}

	chain postnat {
		oifname @wanif udp dport 443 ct original packets 1-9 ip daddr != @nozapret meta mark set meta mark | 0x20000000 queue flags bypass to 200
		oifname @wanif tcp dport { 80, 443 } ct original packets 1-9 ip daddr != @nozapret meta mark set meta mark | 0x20000000 queue flags bypass to 200
	}

	chain postnat_hook {
		type filter hook postrouting priority srcnat + 1; policy accept;
		meta mark & 0x40000000 == 0x00000000 jump postnat
	}

	chain prerouting {
		type filter hook prerouting priority dstnat + 1; policy accept;
	}

	chain prenat {
		type filter hook prerouting priority dstnat - 1; policy accept;
		iifname @wanif tcp sport { 80, 443 } ct reply packets 1-3 ip saddr != @nozapret queue flags bypass to 200
	}

	chain predefrag {
		type filter hook output priority -401; policy accept;
		meta mark & 0x40000000 != 0x00000000 jump predefrag_nfqws comment "nfqws generated : avoid drop by INVALID conntrack state"
	}

	chain predefrag_nfqws {
		meta mark & 0x20000000 != 0x00000000 notrack comment "postnat traffic"
		ip frag-off != 0x0 notrack comment "ipfrag"
		exthdr frag exists notrack comment "ipfrag"
		tcp flags ! syn,rst,ack notrack comment "datanoack"
	}
}
```

/etc/init.d/zapret list_ifsets
```
table inet zapret {
	set lanif {
		type ifname
		elements = { "eth0" }
	}
}
table inet zapret {
	set wanif {
		type ifname
		elements = { "eth0" }
	}
}
table inet zapret {
	set wanif6 {
		type ifname
	}
}
table inet zapret {
	map link_local {
		type ifname : ipv6_addr
	}
}
```

/etc/config/firewall (zones) :
```
config defaults
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	option flow_offloading '0'
	option flow_offloading_hw '1'

config zone
	option name 'lan'
	option input 'ACCEPT'
	option output 'ACCEPT'
	option forward 'ACCEPT'
	list network 'lan'

config zone
	option name 'wan'
	option input 'REJECT'
	option output 'ACCEPT'
	option forward 'REJECT'
	option masq '1'
	option mtu_fix '1'
	list network 'wan'
	list network 'wan6'

config forwarding
	option src 'lan'
	option dest 'wan'
```

**Reply by bol-van** [reply]
i doubt your setup is correct
you created 2 netifd interfaces for single device eth0

what will work for sure :

1. 2 separate adapters
2. multiple vlans. possibly one untagged for clients and one tagged for inet uplink
3. setup second openwrt as default gateway on clients. use masquerade on second openwrt. disable icmp redirects on second openwrt

**Reply by hamedsbt** [reply]
It works with routing mode (not bridge) with one network adapter.

```
sysctl net.ipv4.ip_forward=1
sysctl net.netfilter.nf_conntrack_tcp_be_liberal=1
```

/etc/config/network :
```
config interface 'lan'
	option device 'eth0'
	option proto 'static'
	option ipaddr '192.168.0.1'
	option netmask '255.255.255.0'

config interface 'wan'
	option device 'eth0'
	option proto 'static'
	option ipaddr '192.168.1.2'
	option netmask '255.255.255.0'
	option gateway '192.168.1.1'
```

**Reply by bol-van** [reply]
zapret works fine with one adapter for output
it has problems with forward
it happens because of linux network infrastructure
it's highly optimized and does not always call netfilter
and nfqueue rules are not working

also can be problems with conntrack

this case requires deeper learning of linux internals

**Reply by bol-van** [reply]
first make sure nfqueue is actually not working
run nfqws manually with --debug parameter
then try rules without connbytes (ct) filters
abandon zapret scripts. delete zapret table, kill nfqws
use manual rules. see readme for them

**Reply by bol-van** [reply]
postnat chain should be hook-less

```
nft list chain inet zapret postnat
table inet zapret {
        chain postnat {
                oifname @wanif udp dport 443 ct original packets 1-9 ip daddr != @nozapret meta mark set meta mark | 0x20000000 queue flags bypass to 200
                oifname @wanif tcp dport { 80, 443 } ct original packets 1-9 ip daddr != @nozapret meta mark set meta mark | 0x20000000 queue flags bypass to 200
        }
```

I suggest not altering zapret tables. See "nftables for nfqws" chapter in readme. This is a good start point for testing.

**Reply by KweezyCode** [reply]
есть ли окончательный фикс проблемы? нет возможности поставить zapret на роутер и хотелось бы использовать внешнее устройство как шлюз с zapret

**Comment by hamedsbt** [comment]


**Reply by ValdikSS** [reply]
I didn't use zapret nftables rules, only the single rule I made from the top post. It works for me.
